src/main/java/com/jfrog/ide/common/scan/GraphScanLogic.java

```diff
@@ -0,0 +1,215 @@
+                DependencyTree childCopy = (DependencyTree) child.clone();
+                reducedTree.add(childCopy);
+            }
+        }
+        return reducedTree;
+    }
+
+    @Override
+    public Artifact getArtifactSummary(String componentId) {
+        return scanCache.get(componentId);
+    }
+
+    public static boolean isSupportedInXrayVersion(Version xrayVersion) {
+        return xrayVersion.isAtLeast(MINIMAL_XRAY_VERSION_SUPPORTED_FOR_GRAPH_SCAN);
+    }
+
+    private boolean isSupportedInXrayVersion(Xray xrayClient) {
+        try {
+            if (isSupportedInXrayVersion(xrayClient.system().version())) {
+                return true;
+            }
+            log.error("Unsupported JFrog Xray version: Required JFrog Xray version " + MINIMAL_XRAY_VERSION_SUPPORTED_FOR_GRAPH_SCAN + " and above.");
+        } catch (IOException e) {
+            log.error("JFrog Xray Scan failed. Please check your credentials.", e);
+        }
+        return false;
+    }
+
+    /**
+     * Xray scan a graph of components.
+     *
+     * @param xrayClient      - The Xray client.
+     * @param artifactsToScan - The bulk of components to scan.
+     * @param project         - The JFrog platform project-key parameter to be sent to Xray as context.
+     * @throws IOException in case of connection issues.
+     */
+    private void scanComponents(Xray xrayClient, DependencyTree artifactsToScan, String project) throws IOException {
+        try {
+            if (project != null && !project.isEmpty()) {
+                scanComponentsWithContext(xrayClient, artifactsToScan, project);
+            } else {
+                scanComponentsWithoutContext(xrayClient, artifactsToScan);
+            }
+            // Add to cache non-vulnerable direct dependencies in order to mark them as scanned successfully.
+            // This will allow us to avoid unnecessary future scans.
+            for (DependencyTree child : artifactsToScan.getChildren()) {
+                if (!scanCache.contains(child.getComponentId())) {
+                    scanCache.add(new Artifact(new GeneralInfo(child.toString(),"","",""), new HashSet<>(), new HashSet<>()));
+                }
+            }
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Xray scan a graph of components.
+     * a scan without supplying a context, meaning no specific Xray's watches will be triggered.
+     * only general vulnerabilities and licenses info will be returned (if found).
+     *
+     * @param xrayClient      - The Xray client.
+     * @param artifactsToScan - The bulk of components to scan.
+     * @throws IOException          in case of connection issues.
+     * @throws InterruptedException in case of scan canceled.
+     */
+    private void scanComponentsWithoutContext(Xray xrayClient, DependencyTree artifactsToScan) throws IOException, InterruptedException {
```

Comment:
Looks like there aren't any `checkCanceled.run()` method calls. If they are missing it will be impossible to cancel the scanning. Maybe we should think about how to pass a "check cancelled" method to the Xray client.